### PR TITLE
Add an example for `Encoding` and `Decoding` abilities

### PR DESCRIPTION
**Pull request by lukewilliamboswell**

### CODE REVIEW

**Comment by faldor20** [comment]
Hey, I am slightly concerned that this shows a very inefficient way of decoding and encoding this type. I appreciate you may want to show how the internals work, but it's not really showing best practices. perhaps if we want to show complex encoding and decoding we should pick a type that *should* be decoded that way

eg: For the encoder it should probably look like this:
```roc
encodeItems : ItemKind -> Encoder fmt
encodeItems = \@ItemKind tag ->
    when tag is
        Text -> 1
        Method -> 2
        Function -> 3
        Constructor -> 4
        Field -> 5
        Variable -> 6
        Class -> 7
        Interface -> 8
        Module -> 9
    |>Encode.u8

```

File: examples/EncodeDecode/main.roc [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+ItemKind := [
+    Text,
+    Method,
+    Function,
+    Constructor,
+    Field,
+    Variable,
+    Class,
+    Interface,
+    Module,
```
Comment (bhansconnect) [inline examples/EncodeDecode/main.roc:20]: Would be really cool to make this have 10 values to force at least one to decode 2 digits

File: examples/EncodeDecode/main.roc [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+]
+    implements [
+        Decoding { decoder: decodeItems },
+        Encoding { toEncoder: encodeItems },
+        Inspect,
+        Eq,
+    ]
+
+decodeItems : Decoder ItemKind fmt
+decodeItems = Decode.custom \bytes, _ ->
```
Comment (bhansconnect) [inline examples/EncodeDecode/main.roc:30]: I do like @faldor20 suggestion to do this through encode/decode u8.

That said, feel free to use a custom text encoding if that is preferred to showcase this form of decoding.

File: examples/EncodeDecode/main.roc [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        ['1', .. as rest] -> ok Text rest
+        ['2', .. as rest] -> ok Method rest
+        ['3', .. as rest] -> ok Function rest
+        ['4', .. as rest] -> ok Constructor rest
+        ['5', .. as rest] -> ok Field rest
+        ['6', .. as rest] -> ok Variable rest
+        ['7', .. as rest] -> ok Class rest
+        ['8', .. as rest] -> ok Interface rest
+        ['9', .. as rest] -> ok Module rest
+        _ -> { result: Err TooShort, rest: bytes }
```
Comment (bhansconnect) [inline examples/EncodeDecode/main.roc:46]: Do we have the ability to give other errors? For example if it tries to decode "a". That isn't too short. It just doesn't match at all.

**Review by bhansconnect — APPROVED** [review]
I think it is totally fine to merge as is, but some listed nit comments would be nice to address.

File: examples/EncodeDecode/main.roc (hunk 2 above)
Comment (lukewilliamboswell) [reply examples/EncodeDecode/main.roc]: Done

File: examples/EncodeDecode/main.roc (hunk 3 above)
Comment (lukewilliamboswell) [reply examples/EncodeDecode/main.roc]: No we don't, `TooShort` is the only error available for now in the builtins.

**Comment by lukewilliamboswell** [comment]
Updated with @faldor20 changes, thanks ❤️ 🇦🇺 🦘 . I think this is a much improved example.

**Review by bhansconnect — APPROVED** [review]

**Pull request merged**